src/main/java/server/ConnectionHandler.java

```diff
@@ -0,0 +1,45 @@
+package server;
+
+import server.request.Request;
+import server.request.Parser;
+import server.response.Formatter;
+import server.response.Response;
+import server.response.Selector;
+
+import java.io.IOException;
+
+public class ConnectionHandler implements Runnable {
+    private Client client;
+    private Parser parser;
+    private Selector selector;
+    private Formatter formatter;
+
+    public ConnectionHandler(Client client, Parser parser, Selector selector, Formatter formatter) {
+        this.client = client;
+        this.parser = parser;
+        this.selector = selector;
+        this.formatter = formatter;
+    }
+
+    public void run() {
+        Request request;
+        Response response;
+        try {
+            request = parser.parse();
+            response = selector.selectResponse(request);
+        } catch (IOException err) {
+            response = selector.selectResponse(err);
+        }
+        catch (NullPointerException err) {
```

Comment:
Where does this `NullPointerException` come from? We should be able to handle it without causing an exception if we can anticipate it. https://web.archive.org/web/20140430044213/http://c2.com/cgi-bin/wiki?DontUseExceptionsForFlowControl

Reply:
`parse()` throws a `NullPointerException` if it receives an improperly formatted request (i.e. Chrome sends an empty request, so `parseRequestMethod()` tries to call `split()` on a null `requestLine`).

Reply:
But I supposed that pattern is what that post says is a violation of the principle of least astonishment?

Reply:
Would it be more appropriate to do something like this, since `parseRequestMethod()` is where the exception originates?
```    private String parseRequestMethod(String requestLine) {
        try {
            return requestLine.split(" ")[0].trim();
        } catch (NullPointerException err) {
            return "";
        }
    }```

Reply:
Yup, I wouldn't expect to have to handle a NullPointerException when parsing a Request. If something can be null, we should check it and handle it appropriately. I could see maybe an IllegalArgumentException if you were provided null, but in this case the null is coming from how the request parsing has been implemented.

Reply:
Cool, yeah I just refactored so that `parseRequestLine()` won't return `null`, but an empty request will still cause a 400 status.